feat: bump amp to v3

Yokann opened the pull request:
AMPHP use now Fiber via Revolt eventloop. 
I made this PR to update the current AMPHP implementation of Tornado to this new version. 
I don't think it's a good idea to keep both implementation of AMP (2.6 and 3.0)

- Drop of the PHP 7.4 version was needed.   
- I also update PHPStan version 

## Documentaion
https://amphp.org/upgrade
https://revolt.run/fundamentals

Yokann committed:
feat: bump amp to v3

Yokann committed:
style: update cs-fixer

valentin-claras commented on src/Adapter/Amp/Internal/PromiseWrapper.php:
```diff
@@ -11,11 +11,9 @@
 {
-    /** @var \Amp\Promise */
-    private $ampPromise;
+    private \Amp\Future $ampPromise;
```
Could be readonly now, and the getter removed.

antten replied:
Done ✅

Yokann committed:
test: make promiseAll really async

Yokann commented on src/Adapter/Amp/Internal/Deferred.php:
```diff
@@ -10,13 +10,10 @@
 {
-    /** @var \Amp\Deferred */
-    private $ampDeferred;
+    private \Amp\DeferredFuture $ampDeferred;
+    private PromiseWrapper $promise;
```
What are the pros of making it read-only? It's a private attribute.

antten replied:
It was changed to `private readonly` in this PR: https://github.com/BedrockStreaming/Tornado/pull/53

antten commented on src/Adapter/Amp/EventLoop.php:
```diff
@@ -125,84 +129,83 @@ public function promiseRace(Promise ...$promises): Promise
 
-        \Amp\Loop::delay($milliseconds, function () use ($deferred) {
-            $deferred->resolve();
+        \Revolt\EventLoop::delay($milliseconds / 1000, function () use ($deferred) {
```
**Question**: Why do you now need to convert it into seconds?

Yokann replied:
Because documentation said it takes second.

antten commented on src/Adapter/Amp/EventLoop.php:
```diff
@@ -222,42 +224,41 @@ public function deferred(): Deferred
 
-    /** @var Common\Internal\FailingPromiseCollection */
-    private $unhandledFailingPromises;
+    private Common\Internal\FailingPromiseCollection $unhandledFailingPromises;
```
**Nitpicking**: internal property should be declared at the top of the class.

antten commented on tests/Adapter/Amp/LoopReset.php:
```diff
@@ -13,7 +13,5 @@ class LoopReset implements TestListener
 
     public function endTest(Test $test, float $time): void
```
Why do you keep an empty function? I think you can remove it 🧹

e-zannelli replied:
Was part of the interface, this class was removed on master. The branch need to be rebased on master.

antten commented on tests/EventLoopTest/AsyncTest.php:
```diff
@@ -13,6 +13,7 @@ public function testDummyGenerator(): void
         $generator = (function ($value): \Generator {
+            // @phpstan-ignore-next-line
             return $value;
             yield;  // Mandatory if we want to create a generator
```
**Question**: Why the yield is after the return?

e-zannelli replied:
The comment explains why.

antten commented on tests/EventLoopTest/StreamsTest.php:
```diff
@@ -46,7 +46,7 @@ public function testStreamShouldReadFromWritable(string $expectedSequence = 'W0R
                 // Write twice slower
                 yield $eventLoop->idle();
-                //yield $eventLoop->idle();
+                // yield $eventLoop->idle();
```
I think you can remove this comment.

antten commented on .github/workflows/ci.yml:
```diff
@@ -13,7 +13,7 @@ jobs:
         strategy:
             matrix:
-                php: [ '7.3', '7.4', '8.0' ]
+                php: [ '8.1' ]
```
**Nitpicking**: You can add PHP versions until 8.4.

Yokann replied:
You reviewed an 3 years old PR. 😅Feel free to complete this WIP PR.

e-zannelli replied:
@antten , we need to take ownership of the PR if we want to resurrect it. Maybe even open another PR.

Commit:
following rebase

Commit:
Nitpicking: use import

Commit:
Fix unit tests

Commit:
CS-Fix

Commit:
Execute CI on PHP 8.4

Commit:
Deferred: Remove unused and redondant getPromiseWrapper

Commit:
Replace private property and getter by public readonly property for ampFuture

Commit:
Nitpicking:
- Remove useless comment
- Move property declaration and constructor at the top of the class
- Remove useless index declaration

Commit:
Keep original order on promiseAll

e-zannelli commented on src/Adapter/Amp/EventLoop.php:
```diff
@@ -70,37 +78,48 @@ public function async(\Generator $generator): Promise
+
+            if (count($errors) > 0) {
+                throw reset($errors);
+            }
```
issue(minor,non-blocking): I could be wrong but it seems to me we don't have to wait everything if something fails, we can just use `await` which is like awaitAll without collecting the errors and the first thing that throw will bubble immediately, that seem coherent with other implementations.

```suggestion
            $values = \Amp\Future\await($futures);
```

antten replied:
Oh yes! It's exactly the same implementation between `await` and `awaitAll`, except that `awaitAll` don't stop on exception. Thanks 👍

Commit:
Review fix on promiseAll
